### DISCUSSION TITLE
How to attach DAG_ID, TASK_ID or RUN_ID to my task's logs

### DISCUSSION THREAD

**Discussion by MattiaGallegati** [apache/airflow#24633] [Q&A]
In my recent Airflow exploration, I'm trying to dump logs from Airflow to external systems, possibly without using FluentD, the target system is Splunk in my case.
I was wondering if there is a way to add metainformation like DAG_ID, TASK_ID or RUN_ID to my task's logs in order to provide them with an internal index.
I can see a AIRFLOW__LOGGING__LOG_FORMAT env.var exists, btw it seems like you cannot use Airflow internal variable in this scope but only the Python logger ones.
Can anyone suggest a way to achieve that?

**Comment by tzinckgraf** [comment]
I was able to solve this by adding some custom logging. The default handler is the [FileTaskHandler](https://github.com/apache/airflow/blob/main/airflow/utils/log/file_task_handler.py). You can see the `set_context` function in the `FileTaskHandler` gets a `ti: TaskInstance`. We can use that instance and pass it down to a custom formatter. [This](https://stackoverflow.com/questions/53016338/format-airflow-logs-in-json) is a good stackoverflow link to describe how to create a custom handler.

For the formatter, you can overwrite the formatter by creating a new formatter and adding the class to the `airflow.cfg`. You want to look for the `log_formatter_class = ` option under the `[logging]` section.

I am not sure exactly what type of logging you want to use this for, but this should give you a start. I was able to get this working to add those variables into json logs for something like DataDog.

**Reply by DevilShubert** [reply]
tks a lot
